File: Paths.fs
module Archivist.Paths

open System
open System.IO
open Archivist.Domain

let private nonEmptyOr fallback (value: string) =
    if String.IsNullOrWhiteSpace value then fallback else value

let appName = "archivist"

let private envVar (name: string) =
    Environment.GetEnvironmentVariable(name)
    |> fun value ->
        if String.IsNullOrWhiteSpace value then None
        else Some value

let userHomeDirectory () =
    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)
    |> nonEmptyOr "."

let defaultYoutubeDir () =
    Path.Combine(userHomeDirectory (), "Videos", "YouTube")

let defaultPodcastDir () =
    Path.Combine(userHomeDirectory (), "Music", "Podcasts")

let configDirectory () =
    match envVar "ARCHIVIST_CONFIG_DIR" with
    | Some directory -> directory
    | None ->
        Path.Combine(userHomeDirectory (), ".config")
        |> fun root -> Path.Combine(root, appName)

let configFile () =
    Path.Combine(configDirectory (), "config.json")

let logsDirectory () =
    Path.Combine(configDirectory (), "logs")

let youtubeArchiveDirectory (config: Config) =
    config.youtubeDir

let youtubeArchiveFile (config: Config) (label: string) =
    Path.Combine(youtubeArchiveDirectory config, label, ".download-archive.txt")

let podcastArchiveTemplate (config: Config) =
    Path.Combine(config.podcastDir, "{{podcast_title}}", "archive.json")

File: src/paths.rs
use crate::types::Config;
use crate::util::sanitize_label;
use chrono::Local;
use std::env;
use std::path::{Path, PathBuf};

const APP_NAME: &str = "archivist";

pub fn config_directory() -> PathBuf {
    match env::var_os("ARCHIVIST_CONFIG_DIR") {
        Some(value) if !value.is_empty() => PathBuf::from(value),
        _ => user_home_directory().join(".config").join(APP_NAME),
    }
}

pub fn config_file() -> PathBuf {
    config_directory().join("config.toml")
}

pub fn logs_directory() -> PathBuf {
    config_directory().join("logs")
}

pub fn youtube_archive_file(config: &Config, label: &str) -> PathBuf {
    Path::new(&config.youtube_dir)
        .join(label)
        .join(".download-archive.txt")
}

pub fn podcast_archive_file(config: &Config, label: &str) -> PathBuf {
    Path::new(&config.podcast_dir)
        .join(label)
        .join("archive.json")
}

pub fn sync_log_file(label: &str) -> PathBuf {
    logs_directory().join(format!(
        "sync-{}-{}.log",
        sanitize_label(label),
        Local::now().format("%Y%m%d-%H%M%S")
    ))
}

fn user_home_directory() -> PathBuf {
    env::home_dir()
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}
